addons/mail/tests/discuss/test_bus_presence.py

```diff
@@ -0,0 +1,106 @@
+        )
+        # Guest should not receive im_status of patrick since they don't have a
+        # common channel.
+        self.env["bus.presence"].create({"user_id": patrick.user_ids[0].id, "status": "online"})
+        self.trigger_notification_dispatching([(patrick.partner_id, "presence")])
+        with self.assertRaises(ws._exceptions.WebSocketTimeoutException):
+            websocket.recv()
+        # Guest should receive im_status of bob since they have a common channel.
+        self.env["bus.presence"].create({"user_id": bob.user_ids[0].id, "status": "online"})
+        self.trigger_notification_dispatching([(bob.partner_id, "presence")])
+        notifications = json.loads(websocket.recv())
+        self.assertEqual(notifications[0]["message"]["type"], "bus.bus/im_status_updated")
+        self.assertEqual(notifications[0]["message"]["payload"]["im_status"], "online")
+        self.assertEqual(notifications[0]["message"]["payload"]["partner_id"], bob.partner_id.id)
+
+    def test_user_presence_as_portal_only_shared_channels(self):
+        portal = new_test_user(self.env, login="portal_user", groups="base.group_portal")
+        session = self.authenticate("portal_user", "portal_user")
+        bob = new_test_user(self.env, login="bob_user", groups="base.group_user")
+        patrick = new_test_user(self.env, login="patrick_user", groups="base.group_user")
+        channel = self.env["discuss.channel"].channel_create(group_id=None, name="General")
+        channel.add_members(partner_ids=[portal.partner_id.id, bob.partner_id.id])
+        websocket = self.websocket_connect(cookie=f"session_id={session.sid};")
+        self.subscribe(
+            websocket,
+            [
+                f"odoo-presence-res.partner_{bob.partner_id.id}",
+                f"odoo-presence-res.partner_{patrick.partner_id.id}",
+            ],
+            self.env["bus.bus"]._bus_last_id(),
+        )
+        # Portal should not receive im_status of patrick since they don't have a
+        # common channel.
+        self.env["bus.presence"].create({"user_id": patrick.user_ids[0].id, "status": "online"})
+        self.trigger_notification_dispatching([(patrick.partner_id, "presence")])
+        with self.assertRaises(ws._exceptions.WebSocketTimeoutException):
+            websocket.recv()
+        # Portal should receive im_status of bob since they have a common channel.
+        self.env["bus.presence"].create({"user_id": bob.user_ids[0].id, "status": "online"})
+        self.trigger_notification_dispatching([(bob.partner_id, "presence")])
+        notifications = json.loads(websocket.recv())
+        self.assertEqual(notifications[0]["message"]["type"], "bus.bus/im_status_updated")
+        self.assertEqual(notifications[0]["message"]["payload"]["im_status"], "online")
+        self.assertEqual(notifications[0]["message"]["payload"]["partner_id"], bob.partner_id.id)
+
+    def test_guest_presence_as_user_only_shared_channels(self):
```

Comment:
as portal?

Why only 3 tests?

They should be 4 cases (guest requesting user, guest requesting guest, portal requesting user, portal requesting guest)

Reply:
Maybe we could factorize a bit some of the code and only have 1 or 2 combined tests

Reply:
Actually I think there should be 6 if we want 100% coverage:
- guest receives user, guest receives guest
- portal receives user, portal receives guest
- internal receives user, internal receives guest
I grouped those scenarios (guest, portal, internal) into 3 tests which use a common helper.